### DISCUSSION TITLE
mimic esbuild module transformer

### DISCUSSION THREAD

**Discussion by magic-akari** [swc-project/swc#3940] [General]
Giving the following input

```javascript
export let foo = 1;

export const inc = () => {
    return foo++;
}

export function dec(){
    return foo--;
}
```

[esbuild output](https://esbuild.egoist.sh/#W1siaW5kZXgudHMiLHsiY29udGVudCI6ImV4cG9ydCBsZXQgZm9vID0gMTtcblxuZXhwb3J0IGNvbnN0IGluYyA9ICgpID0+IHtcbiAgcmV0dXJuIGZvbysrO1xufVxuXG5leHBvcnQgZnVuY3Rpb24gZGVjKCl7XG4gIHJldHVybiBmb28tLTtcbn0ifV0sWyJlc2J1aWxkLmNvbmZpZy5qc29uIix7ImNvbnRlbnQiOiJ7XG4gIFwiZm9ybWF0XCI6IFwiY2pzXCIsXG4gIFwiY2RuVXJsXCI6IFwiaHR0cHM6Ly9jZG4uc2t5cGFjay5kZXZcIlxufSJ9XV0=)(click bundle button to see result)
```javascript
var __defProp = Object.defineProperty;
var __markAsModule = (target) => __defProp(target, "__esModule", { value: true });
var __export = (target, all) => {
  __markAsModule(target);
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};

__export(exports, {
  dec: () => dec,
  foo: () => foo,
  inc: () => inc
});
var foo = 1;
var inc = () => {
  return foo++;
};
function dec() {
  return foo--;
}
```

It uses var hoist and value getter to achieve a cleaner result.
It does not require manually sync variable assignment to `exports`.
I don't know if it is any different from spec behavior.

**Comment by kdy1** [comment]
I love the idea, but we should check the specification

**Comment by magic-akari** [comment]
Some useful information
- https://sokra.github.io/interop-test/
